### ISSUE TITLE
[ENH] A variation of CutoffSplitter that allows for the options of (1) using all training data possible and (2) varying forecast horizons for each split

### ISSUE THREAD

**Issue by GianlucaGuarroSonos** [sktime/sktime#3682]
**Is your feature request related to a problem? Please describe.**
Right now the CutoffSplitter splits time series data at given cutoff points into fixed-length training and test sets.
I believe it would be useful to extend the functionality of the CutoffSplitter to allow it to have the following two features:

1. The training sets of the splits can be composed of ALL the data prior to the cutoff instead of a fixed amount. (The same idea behind ExpandingWindowSplitter).
2. The test sets of each split can vary in size. The motivation behind this is for when I am building models at a disaggregate time series level but I care about the performance at a more aggregate level in which the sizes of the test sets at the disaggregate level are not necessarily the same. For example consider that I train my model on weekly data but I want to test the performance using a year's worth of data from the cutoff point. Depending on the year this may be 52 weeks or 53 weeks so I cannot use a fixed length forecasting horizon for each split.

**Describe the solution you'd like**
**For feature 1**
I am not sure whether these functionalities should be built into the existing CutoffSplitter or if a different splitter should be created. In any case if we wanted to use the existing CutoffSplitter, in order to satisfy feature (1), the user could pass in a "-1" to `window_length` to indicates to use the expanding window lengths. Alternatively a new boolean parameter can be created called `use_expanding` to indicate whether the windows should be expanding or sliding.

**For feature 2**
Perhaps the easiest and most flexible solution would be to add another parameter that allows for a list of forecasting horizons or ints which serves to specify how long each test set should be in each window. The length of this array should naturally be the same as the length of the `cutoffs` array.

**Describe alternatives you've considered**
Alternative solution would be to create an entire new Splitter class altogether to disentangle whether the created windows should be sliding or expanding (thus addressing feature request (1)). This would involve renaming the existing `CutoffSplitter` to something like `CutoffSlidingSplitter` and the creation of a new splitter called `CutoffExpandingSplitter`. Feature request (2) would have to be added to both classes and perhaps the `ExpandingWindowSplitter` and the `SlidingWindowSplitter` as well.

**Comment by fkiraly** [comment]
> The training sets of the splits can be composed of ALL the data prior to the cutoff instead of a fixed amount. (The same idea behind ExpandingWindowSplitter).

Ah, I was under the impression that the `CutoffSplitter` uses all data, but it does not. Either way, if it is used in `update` mode, the effect will be the same as in feature 1.

Design-wise, I would agree with your suggestions:
* `window_length=None` (not -1 which would be strange) should return all the data prior to the cutoff.
* `fh` as a list makes sense, but unfortunately this would imply an ambiguity. Currently, `int` is allowed, so it is ambiguous what list of int should be. So, allow list of any objects as current, unless it creates an ambiguity (so exclude list of int).

Would appreciate a PR!

Any additional thoughts, @khrapovs, as you have looked at the splitters last?

**Comment by GianlucaGuarroSonos** [comment]
> fh as a list makes sense, but unfortunately this would imply an ambiguity. Currently, int is allowed, so it is ambiguous what list of int should be. So, allow list of any objects as current, unless it creates an ambiguity (so exclude list of int).

maybe it would have to be a list of list of ints where each sublist is the time steps you want to predict for a particular split.

> Would appreciate a PR!

I could give this a shot. This would be my first significant PR in any open-source project but seems doable for me.

P.S.
How would you foresee the appropriate way to compare models trained at differing temporal aggregates? Obviously any predictions made on the disaggregate models would have to be aggregated together in order to compare performances of the most aggregate model. Right now, I imagine it can be done by supplying a custom `scoring` method to `ForecastingGridSearchCV` or `ForecastingRandomizedSearchCV` in which the user has to figure out how to group and aggregate predictions together. It's a little weird though because that `scoring` method would have to access global variables. But to get around that, I think we could use [partial functions](https://docs.python.org/3/library/functools.html#functools.partial). 

That was a bit stream of consciousness but curious if you think that would be an appropriate way to handle that issue or if you have other ideas.

**Comment by fkiraly** [comment]
> I could give this a shot. This would be my first significant PR in any open-source project but seems doable for me.

Nice, thanks!

I would suggest to separate the two features and go for maybe feature 1 first?

Contrib guide is here:
https://www.sktime.org/en/latest/get_involved/contributing.html

> How would you foresee the appropriate way to compare models trained at differing temporal aggregates?

That sounds like pears and apples. Would that relate to this feature? Or is it a separate question?

**Comment by GianlucaGuarroSonos** [comment]
> I would suggest to separate the two features and go for maybe feature 1 first?
Yeah splitting it into 2 PR's make sense. I hope to try and tackle feature 1 soon then.

> That sounds like pears and apples. Would that relate to this feature? Or is it a separate question?
It starts off being pears and apples but the core advantage of feature 2 is to let you transform these comparison to apples and apples. In effect, it allows the user to treat the temporal granularity of the time series model as a hyperparameter. For example, say you have 2 models:

- Model A is trained on monthly aggregations of the forecast variable
- Model B is trained on weekly aggregations of the forecast variable

I would like the cross-validation scores of Model B to be on the same scale as that of Model A so I can easily compare them. Clearly the only way to do this is to transform the weekly forecasts of Model B to monthly forecasts by aggregating the appropriate weekly forecasts together for each month. You do the same aggregation of weekly data to monthly for the actuals. And now you can assess Model B's performance as if it was making monthly predictions. Now you can compare Model B's performance with that of Model A's because the metrics are on the same scale of monthly.

So it relates in that feature 2 is a requirement to being able to treat the level of temporal aggregation as a hyperparameter. I'm just not sure if the solution I proposed by using a custom scoring method is the most appropriate. Whatever the solution, I think it would be something nice to be able to do in sktime as exploring different temporal aggregations is something that pops up in the literature from time to time.
